Source/WebCore/rendering/RenderGrid.cpp

```diff
@@ -778,6 +818,47 @@ void RenderGrid::placeItemsOnGrid(std::optional<LayoutUnit> availableLogicalWidt
 #endif
 }
 
+LayoutUnit RenderGrid::masonryContentSize() const
+{
+    return m_masonryLayout.gridContentSize();
+}
+
+static bool itemGridAreaStartsAtFirstLine(const GridArea& area, GridTrackSizingDirection masonryDirection)
+{
+    return !(masonryDirection == ForRows ? area.rows.startLine() : area.columns.startLine());
+}
+
+bool RenderGrid::hasDefiniteGridAxisPosition(const RenderBox* child, GridTrackSizingDirection gridAxisDirection) const 
+{
+    auto itemSpan = GridPositionsResolver::resolveGridPositionsFromStyle(*this, *child, gridAxisDirection);
+    return !itemSpan.isIndefinite();
+}
+
+static bool itemGridAreaIsWithinImplicitGrid(const GridArea& area, unsigned gridAxisLinesCount, GridTrackSizingDirection gridAxisDirection)
+{
+    auto itemSpan = gridAxisDirection == ForColumns ? area.columns : area.rows;
+    return itemSpan.startLine() <  gridAxisLinesCount && itemSpan.endLine() < gridAxisLinesCount;
+}
+
+void RenderGrid::collectMasonryItems(Grid& grid, unsigned availableGridAxisTracks, GridTrackSizingDirection masonryDirection, HashMap<RenderBox*, GridArea>& itemsOnFirstTrack, Vector<RenderBox*>& itemsWithDefiniteGridAxisPosition, Vector<RenderBox*>& itemsWithIndefinitePosition) const
```

Comment:
Since the `itemsOnFirstTrack `, `itemsWithDefiniteGridAxisPosition `, and `itemsWithIndefinitePosition ` arguments enter as empty Vectors, it might be better to reserve space for each before entering the loop, perhaps by using the grid width/height dimensions or something as a rough guide. That would avoid having to realloc the Vector (potentially multiple times) if you exceed it's default resize step.